src/bonefish/router/wamp_router_impl.cpp

```diff
@@ -179,19 +189,70 @@ void wamp_router_impl::process_hello_message(const wamp_session_id& session_id,
+            }
+        }
 
-    // If we fail to send the welcome message it is most likely that the
-    // underlying network connection has been closed/lost which means
-    // that the component is no longer reachable on this session. So all
-    // we do here is trace the fact that this event occured.
-    BONEFISH_TRACE("%1%, %2%", *session % *welcome_message);
-    if (!session->get_transport()->send_message(std::move(*welcome_message))) {
-        BONEFISH_ERROR("failed to send the welcome message: network failure");
+        std::optional<wamp_authenticator::challenge> challenge {
+            m_authenticator->generate_challenge(hello_details.get_auth_id(), selected_auth_method)
+        };
+
+        if (challenge) {
+            m_session_auth_info.insert_or_assign(session_id,
+                                                 wamp_auth_info{challenge->auth.id, challenge->auth.method});
+
+            std::unique_ptr<wamp_challenge_message> challenge_message(new wamp_challenge_message);
+            challenge_message->set_auth_method(challenge->auth.method);
+            challenge_message->set_extra(challenge->extra.marshal(challenge_message->get_zone()));
+
+            session->set_state(wamp_session_state::CHALLENGING);
+
+            BONEFISH_TRACE("%1%, %2%", *session % *challenge_message);
+            if (!session->get_transport()->send_message(std::move(*challenge_message))) {
+                BONEFISH_TRACE("failed to send the challenge message: network failure");
+            }
+        } else {
+            // TODO: add message to abort_message, see examples at
+            // https://wamp-proto.org/wamp_latest_ietf.html#section-4.1.3-12
+            std::unique_ptr<wamp_abort_message> abort_message(new wamp_abort_message);
+            abort_message->set_reason("wamp.error.not_authorized");
+            BONEFISH_TRACE("%1%, %2%, %3%", *session % *hello_message % *abort_message);
+            if (!session->get_transport()->send_message(std::move(*abort_message))) {
+                BONEFISH_TRACE("failed to send the welcome message: network failure");
+            }
+        }
+    } else {
+        session->set_state(wamp_session_state::OPEN);
+
+        std::unique_ptr<wamp_welcome_message> welcome_message(new wamp_welcome_message);
+        welcome_message->set_session_id(session_id);
+        welcome_message->set_details(m_welcome_details.marshal(welcome_message->get_zone()));
+
+        // If we fail to send the welcome message it is most likely that the
+        // underlying network connection has been closed/lost which means
+        // that the component is no longer reachable on this session. So all
+        // we do here is trace the fact that this event occured.
+        BONEFISH_TRACE("%1%, %2%", *session % *welcome_message);
+        if (!session->get_transport()->send_message(std::move(*welcome_message))) {
+            BONEFISH_TRACE("failed to send the welcome message: network failure");
```

Comment:
`BONEFISH_ERROR` is changed to `BONEFISH_TRACE`, is it intended?

Also, in above cases errors are logged by `BONEFISH_TRACE`.

Reply:
done